HelloWorld_ST_Sensors/user/gps/gps.cpp

```diff
@@ -0,0 +1,152 @@
+                        break;
+                    }
+                    break;
+                case GPS_GPRMC:
+                    switch (nmea_string_token_index)
+                    {
+                    case GPRMC_VALIDITY_INDEX:
+                        strcpy(gps_coordinates_data.validity, temp_token);
+                        // Consider NMEA strings that are GPRMC and Active Navigation Receiver Warning
+                        if (strcmp(temp_token, "A") != 0)
+                        {
+                            gps_nmea_type = GPS_INVALID_NMEA;
+                        }
+                        break;
+                    case GPRMC_LATITUDE_INDEX:
+                        strcpy(gps_coordinates_data.latitude, temp_token);
+                        break;
+                    case GPRMC_LATITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.lat_dir, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_INDEX:
+                        strcpy(gps_coordinates_data.longitude, temp_token);
+                        break;
+                    case GPRMC_LONGITUDE_DIR_INDEX:
+                        strcpy(gps_coordinates_data.long_dir, temp_token);
+                        break;
+                    }
+                    break;
+                }
+            }
+        }
+        return gps_nmea_type;
+    }
+
+    uint8_t gps_c::extract_gpgga_or_gprmc_line(char *gps_buff)
+    {
+        uint8_t gps_nmea_type = GPS_INVALID_NMEA;
+        char valid_gps_string[100] = {0};
+        uint16_t counter = 0;
+        for (uint16_t i = 0; i < UART_BUF_LEN; i++)
+        {
+            if (gps_buff[i] == '$')
+            {
+                while (gps_buff[i] != '\n' && i < UART_BUF_LEN)
+                {
+                    valid_gps_string[counter] = gps_buff[i];
+                    i++;
+                    counter++;
+                }
+                if (i < UART_BUF_LEN)
+                {
+                    gps_nmea_type = check_gprmc_gpgga_line(valid_gps_string);
+                    if (gps_nmea_type != GPS_INVALID_NMEA)
+                    {
+                        break;
+                    }
+                }
+                counter = 0;
+                memset(valid_gps_string, 0, strlen(valid_gps_string));
+            }
+        }
+        return gps_nmea_type;
+    }
+
+    void gps_c::register_func(void (*external_fn)(const char *newline))
+    {
+        gps_process_fn_ptr = external_fn;
+    }
+
+    void gps_c::gps_interrupt_cb()
+    {
+        int data = gps_uart.getc();
+        curr_line_buffer[index] = data;
+        if (index != UART_BUF_LEN)
+        {
+            index++;
+        }
+        if (index == UART_BUF_LEN)
+        {
+            uint8_t gps_nmea_type = extract_gpgga_or_gprmc_line(curr_line_buffer);
+            if ((GPS_INVALID_NMEA != gps_nmea_type) && (gps_process_fn_ptr != NULL))
```

Comment:
I can see that you have handled the case where gps_nmea_type is not valid for our purposes
And you are calling the callback function here. 

But there isn't any parsing logic for the gpgga / gprmc in the else clause.

Reply:
Okay I realize you might be update the gpgga/gprmc in the `extract_gpgga_or_gprmc_line` function.
Please make the functions concise and limit their scope to one action